rfcs/text/0015_nodejs_clustering.md

```diff
@@ -0,0 +1,509 @@
+enhance the log record with context information before forwarding it to its appenders, but this still
+seems to be the best solution.
+
+### The rolling-file appender
+
+The rolling process of the `rolling-file` appender is going to be problematic in clustered mode, as it will cause 
+concurrency issues during the rolling. We need to find a way to have this rolling process clustered-proof.
+
+Identified options are:
+
+- have the rolling file appenders coordinate themselves when rolling
+
+By using a broadcast message based mutex mechanism, the appenders could acquire a ‘lock’ to roll a specific file, and release the lock once the rotation is done.
+
+- have the coordinator process perform the rolling
+
+Another option would be to have the coordinator perform the rotation instead. When a rolling is required, the appender would send a message to the coordinator, which would perform the rolling and notify the workers once the process is complete. Note that this option is even more complicated than the previous one, as it forces to move the rolling implementation outside of the appender, without any significant upsides.
+
+- centralize the logging system in the coordinator (and have the workers send messages to the coordinator when using the logging system)
+
+We could go further, and change the way the logging system works in clustering mode by having the coordinator centralize the logging system. The worker’s logger implementation would just send messages to the coordinator. If this may be a correct design, the main downside is that the logging implementation would be totally different in cluster and non cluster mode, and is overall way more work that the other options.
+
+Overall, if no option is trivial, I feel like option 1) is still the most pragmatic one. Option 3) is probably a better 
+design, but represents more work. It would probably be fine if all our appenders were impacted, but as only the 
+rolling-file one is, I feel like going with 1) would be ok.
+
+### The status API
+
+In clustering mode, the workers will all have an individual status. One could have a connectivity issue with ES 
+while the other ones are green. Hitting the `/status` endpoint will reach a random (and different each time) worker, 
+meaning that it would not be possible to know the status of the cluster as a whole.
+
+We will need to add a centralized status service in the coordinator. Also, as the `/status` endpoint cannot be served 
+from the coordinator, we will also need to have the workers retrieve the global status from the coordinator to serve 
+the status endpoint.
+
+We may also want to have the `/status` endpoint display each individual worker status in addition to the 
+global status, which may be breaking change in the `/status` API response format.
+
+### PID file
+
+Without changes, each worker is going to try to write and read the same PID file. Also, this breaks the whole pid file 
```

Comment:
What's the purpose of or how does Kibana use the PID file?

Reply:
@jbudz should know better than me, but my understanding is that it's used by our RPM packages for system service shutdowns?

Reply:
:+1: - in 7.x we support SysV, which reads from the PID file for statuses, [ref](https://github.com/elastic/kibana/blob/7f2afac9972f44cbdb7f335b7b94b6af39e891f0/src/dev/build/tasks/os_packages/service_templates/sysv/etc/init.d/kibana#L106).  The file is generated by the startup script (not using the kibana.yml setting).

We've removed SysV in support in 8.0, so this isn't strictly necessary but still useful to have for external process monitors to attach to.  One example is monit, it will monitor cpu, memory usage, and other metrics and send out alerts if a process is misbehaving.